Allow text to be set to a transparency.

megangardner opened the pull request:
#### 8ba4bc78b36ed01e6370d63c7ec8dd2ec3e06140
<pre>
Allow text to be set to a transparency.
<a href="https://bugs.webkit.org/show_bug.cgi?id=273602">https://bugs.webkit.org/show_bug.cgi?id=273602</a>
<a href="https://rdar.apple.com/127309911">rdar://127309911</a>

Reviewed by Wenson Hsieh and Richard Robinson.

We need a way to mark text as transparent, so add
a document marker to set the transparency. This could
be combined with the DraggedContent document markers in the
future.

* Source/WebCore/dom/DocumentMarker.h:
(WebCore::DocumentMarker::allMarkers):
* Source/WebCore/dom/DocumentMarkerController.cpp:
(WebCore::shouldInsertAsSeparateMarker):
* Source/WebCore/rendering/MarkedText.cpp:
(WebCore::MarkedText::collectForDocumentMarkers):
* Source/WebCore/rendering/MarkedText.h:
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::contentContainsReplacedElement):
(WebCore::RenderReplaced::paint):
(WebCore::draggedContentContainsReplacedElement): Deleted.
* Source/WebCore/rendering/StyledMarkedText.cpp:
(WebCore::resolveStyleForMarkedText):
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::TextBoxPainter&lt;TextBoxPath&gt;::paintForeground):
(WebCore::TextBoxPainter&lt;TextBoxPath&gt;::createDecorationPainter):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::getRangeforUUID):
(WebKit::WebPage::getTextIndicatorForID):
(WebKit::WebPage::updateTextIndicatorStyleVisibilityForID):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Tools/TestWebKitAPI/Tests/WebCore/MarkedText.cpp:
(WebCore::operator&lt;&lt;):

Canonical link: <a href="https://commits.webkit.org/278276@main">https://commits.webkit.org/278276@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2c0a59bb6355ba7f223b23b324b23b75a420a5a6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50007 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29296 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2290 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53250 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/684 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52308 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/35434 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/253 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40785 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52105 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26911 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/43098 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21917 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/24342 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/259 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8460 "Built successfully") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/46391 "Build is in progress. Recent messages:OS: Sonoma (14.4.1), Xcode: 15.3; Running apply-patch; Checked out pull request") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/306 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54832 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25101 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/243 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48177 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26359 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43218 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/47293 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27221 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7224 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26088 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -311,14 +311,21 @@
 {
-    auto sessionRange = m_unifiedTextReplacementController->contextRangeForSessionWithUUID(uuid);
-    if (!sessionRange) {
+    auto range = m_unifiedTextReplacementController->contextRangeForSessionWithUUID(uuid);
```
this can probably be inverted to have less nested branches, and I think `makeSimpleRange` null checks the parameter too.

```
auto range = m_unifiedTextReplacementController->contextRangeForSessionWithUUID(uuid);
if (range)
    return range;

RefPtr liveRange = m_textIndicatorStyleEnablementRanges.get(uuid);

return WebCore::makeSimpleRange(liveRange);

rr-codes commented on Source/WebCore/dom/DocumentMarker.h:
```diff
@@ -124,6 +125,9 @@ class DocumentMarker : public CanMakeWeakPtr<DocumentMarker> {
     };
 #endif
+    struct TransparentContentData {
+        float alphaValue { 0.0 };
```
I think just `alpha` is sufficient

whsieh commented on Source/WebCore/dom/DocumentMarker.h:
```diff
@@ -124,6 +125,9 @@ class DocumentMarker : public CanMakeWeakPtr<DocumentMarker> {
 #endif
+    struct TransparentContentData {
+        float alphaValue { 0.0 };
+    };
```
Since this isn't supported yet, I'd just leave it out of this PR altogether. We can reintroduce this once this can become unified with `DraggedContentRanges`.

(Then you can also remove the `FIXME` below)

whsieh commented on Source/WebCore/rendering/TextBoxPainter.cpp:
```diff
@@ -515,8 +515,8 @@ void TextBoxPainter<TextBoxPath>::paintForeground(const StyledMarkedText& marked
-    GraphicsContextStateSaver stateSaver(context, markedText.style.textStyles.strokeWidth > 0 || markedText.type == MarkedText::Type::DraggedContent);
-    if (markedText.type == MarkedText::Type::DraggedContent)
+    GraphicsContextStateSaver stateSaver(context, markedText.style.textStyles.strokeWidth > 0 || markedText.type == MarkedText::Type::DraggedContent || markedText.type == MarkedText::Type::TransparentContent);
+    if (markedText.type == MarkedText::Type::DraggedContent || markedText.type == MarkedText::Type::TransparentContent)
```
Maybe pull this out into a local variable?

```suggestion
    bool isTransparentMarkedText = markedText.type == MarkedText::Type::DraggedContent || markedText.type == MarkedText::Type::TransparentContent;
    GraphicsContextStateSaver stateSaver(context, markedText.style.textStyles.strokeWidth > 0 || isTransparentMarkedText);
    if (isTransparentMarkedText)
```

whsieh commented on Source/WebCore/rendering/RenderReplaced.cpp:
```diff
@@ -246,10 +246,16 @@ void RenderReplaced::paint(PaintInfo& paintInfo, const LayoutPoint& paintOffset)
         auto* parentContainer = element()->parentOrShadowHostElement();
         ASSERT(parentContainer);
-        if (CheckedPtr markers = document().markersIfExists(); markers && draggedContentContainsReplacedElement(markers->markersFor(*parentContainer, DocumentMarker::Type::DraggedContent), *element())) {
+        if (CheckedPtr markers = document().markersIfExists(); markers && contentContainsReplacedElement(markers->markersFor(*parentContainer, DocumentMarker::Type::DraggedContent), *element())) {
```
You could probably make the `markers` null check cover both branches:

```
if (CheckedPtr markers = document().markersIfExists() {
    if (draggedContentContainsReplacedElement(markers->markersFor(*parentContainer, DocumentMarker::Type::DraggedContent), *element()))
        …
    if (contentContainsReplacedElement(markers->markersFor(*parentContainer, DocumentMarker::Type::TransparentContent), *element()))
        …
}
```

megangardner committed:
Allow text to be set to a transparency.
https://bugs.webkit.org/show_bug.cgi?id=273602
rdar://127309911

Reviewed by Wenson Hsieh and Richard Robinson.

We need a way to mark text as transparent, so add
a document marker to set the transparency. This could
be combined with the DraggedContent document markers in the
future.

* Source/WebCore/dom/DocumentMarker.h:
(WebCore::DocumentMarker::allMarkers):
* Source/WebCore/dom/DocumentMarkerController.cpp:
(WebCore::shouldInsertAsSeparateMarker):
* Source/WebCore/rendering/MarkedText.cpp:
(WebCore::MarkedText::collectForDocumentMarkers):
* Source/WebCore/rendering/MarkedText.h:
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::contentContainsReplacedElement):
(WebCore::RenderReplaced::paint):
(WebCore::draggedContentContainsReplacedElement): Deleted.
* Source/WebCore/rendering/StyledMarkedText.cpp:
(WebCore::resolveStyleForMarkedText):
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::TextBoxPainter<TextBoxPath>::paintForeground):
(WebCore::TextBoxPainter<TextBoxPath>::createDecorationPainter):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::getRangeforUUID):
(WebKit::WebPage::getTextIndicatorForID):
(WebKit::WebPage::updateTextIndicatorStyleVisibilityForID):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Tools/TestWebKitAPI/Tests/WebCore/MarkedText.cpp:
(WebCore::operator<<):

Canonical link: https://commits.webkit.org/278276@main
